presto-main/src/main/java/com/facebook/presto/cost/StatisticRange.java

```diff
@@ -160,20 +160,35 @@ public StatisticRange addAndCollapseDistinctValues(StatisticRange other)
     public StatisticRange subtract(StatisticRange rightRange)
     {
         StatisticRange intersect = intersect(rightRange);
-        double newLow = getLow();
-        double newHigh = getHigh();
-        if (intersect.getLow() == getLow()) {
-            newLow = intersect.getHigh();
+
+        double overlapPercentOfThis = overlapPercentWith(rightRange);
+        double overlapDistinctValuesThis = overlapPercentOfThis * distinctValues;
+
+        if (intersect.getDistinctValuesCount() >= overlapDistinctValuesThis) {
+            // all NDVs from intersection are subtracted
```

Comment:
it didn't have any reasonable semantics previously. How it checks how many NDVs are removed from intersecting range and constructs result range accordingly. Sounds like an improvement to what was previously.